### PR TITLE
feat: add support for visualization alternatives

### PR DESCRIPTION
**Pull request by toftalisk**

### CODE REVIEW

File: src/components/components.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { FC, useEffect, useState } from 'react';
+import { GraphNode, GraphEdge, GraphSelection } from '../models/graphModel';
+import { patchGraph } from './state/patchGraph';
+import { GraphProps, GraphStateProps } from './state/GraphStateProps';
+import { RdfContext } from './state/RdfContext';
+import { RdfStateProps } from './state/RdfState.types';
+import { F3DGraph } from './f3dGraph/F3DGraph';
+import { CyGraph } from './cyGraph/CyGraph';
```
Comment (DzianisMakeichyk) [inline src/components/components.tsx:8]: use `index.ts` for it :)

File: src/components/components.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+function createRdfGraphHoc<P extends GraphProps, R = Omit<P, keyof GraphProps>>(Component: FC<P>): FC<R & RdfStateProps> {
+	return ({ rdfStore, rdfPatch, onElementSelected, ...props }: RdfStateProps) => {
+		const forwardSelection = (selection: GraphSelection) => {
+			onElementSelected(selection);
```
Comment (DzianisMakeichyk) [inline src/components/components.tsx:13]: looks like `forwardSekection` no needs

File: src/components/cyGraph/CyGraph.tsx [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+	GraphNodeIdentifier,
+	GraphPatch,
+	GraphPropertyIdentifier,
+} from '../../models/graphModel';
+import { NodeSymbol } from '../../symbol-api';
+import cytoscape from 'cytoscape';
+
+const addNode = (n: GraphNodeIdentifier, cy: Cytoscape.Core) => {
+	const { id, type } = n;
+	const elem: ElementDefinition = { data: { id: id } };
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:23]:
```suggestion
	const elem: ElementDefinition = { data: { id } };
```

**Review by toftalisk — COMMENTED** [review]
Comments after group review

File: src/components/cyGraph/CyGraph.tsx [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+	GraphEdge,
+	GraphEdgeIdentifier,
+	GraphNodeIdentifier,
+	GraphPatch,
+	GraphPropertyIdentifier,
+} from '../../models/graphModel';
+import { NodeSymbol } from '../../symbol-api';
+import cytoscape from 'cytoscape';
+
+const addNode = (n: GraphNodeIdentifier, cy: Cytoscape.Core) => {
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:21]:
```suggestion
const addNode = ({ id, type }: GraphNodeIdentifier, cy: Cytoscape.Core) => {
```

File: src/components/cyGraph/CyGraph.tsx [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+	const { id, type } = n;
+	const elem: ElementDefinition = { data: { id: id } };
+	if (type === 'linkNode') {
+		if (!elem.style) elem.style = {};
+		elem.style.display = 'none';
+	}
+	cy.add(elem);
+};
+
+const addProperty = (prop: GraphPropertyIdentifier, cy: Cytoscape.Core) => {
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:31]:
```suggestion
const addProperty = ({ key, ..... } : GraphPropertyIdentifier, cy: Cytoscape.Core) => {
```

File: src/components/cyGraph/CyGraph.tsx [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+			break;
+		case 'parent':
+			node.move({ parent: prop.node.parent!.id });
+			break;
+		default:
+			node.data(prop.key, prop.value);
+	}
+};
+
+const addEdge = (n: GraphEdge, cy: Cytoscape.Core) => {
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:51]:
```suggestion
const addEdge = ({ id, source, target }: GraphEdge, cy: Cytoscape.Core) => {
```

File: src/components/cyGraph/CyGraph.tsx [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+	const { id, source, target } = n;
+	const { [labelIri]: label } = n.linkRef! || {};
+	const elem: ElementDefinition = { data: { id: id, source: source, target: target, [labelKey]: label } };
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:54]:
```suggestion
	const elem: ElementDefinition = { data: { id, source, target, [labelKey]: label } };
```

File: src/components/cyGraph/CyGraph.tsx [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+	cy.add(elem);
+	return [];
+};
+const removeElement = (element: GraphEdgeIdentifier | GraphNodeIdentifier, cy: Cytoscape.Core) => {
+	cy.remove(cy.getElementById(element.id));
+};
+const removeProperty = (prop: GraphPropertyIdentifier, cy: Cytoscape.Core) => {
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:61]:
```suggestion
const removeProperty = ({ node, key }: GraphPropertyIdentifier, cy: Cytoscape.Core) => {
```

File: src/components/cyGraph/CyGraph.tsx [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+	cy.remove(`[id = "${getImageNodeId(compoundNodeId)}"]`);
+};
+
+const nodeTypeKey = 'nodeType';
+const layoutIgnoreKey = 'layoutIgnore';
+
+const applyPatch = (graphPatch: GraphPatch, cy: Cytoscape.Core) => {
+	const postprocess: (() => void)[] = [];
+
+	for (const a of graphPatch) {
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:97]:
```suggestion
	for (const { action, assertion, ........ } of graphPatch) {
```

File: src/components/cyGraph/CyGraph.tsx [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+	const connectorSelector = `[nodeType = "${NodeType.SymbolConnector}"]`;
+	const imageSelector = `[nodeType = "${NodeType.SymbolImage}"]`;
+
+	_event.cy
+		.$(`${connectorSelector}, ${imageSelector}`)
+		.layout({
+			name: 'preset',
+			animate: false,
+			transform: (node, pos) => {
+				const parentPosition = node.parent().first().position();
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:148]:
```suggestion
				const { x, y } = node.parent().first().position();
```

File: src/components/cyGraph/CyGraph.tsx [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+		}
+		nullableCy.batch(() => {
+			applyPatch(graphPatch, nullableCy);
+		});
+
+		runLayout(nullableCy);
+	}, [graphPatch]);
+
+	useEffect(() => {
+		if (nullableCy) {
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/CyGraph.tsx:182]:
```suggestion
		if (!nullableCy) return;
```

File: src/components/cyGraph/StoryWrapper.tsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
+import { Button } from '@equinor/eds-core-react';
+import { useEffect, useReducer, useState } from 'react';
+import { colorPredicate, hasSvgPredicate } from '../../mapper/predicates';
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/StoryWrapper.tsx:3]: => `index.ts` :)

File: src/components/cyGraph/StoryWrapper.tsx [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+		setNodeNumber((n) => n + 1);
+	};
+
+	const loadTurtle = (): void => {
+		updateTurtle(turtleString);
+	};
+
+	const handleSelection = (selection: GraphSelection) => {
+		setSelection(selection);
+	};
```
Comment (DzianisMakeichyk) [inline src/components/cyGraph/StoryWrapper.tsx:53]: wrapper necessary

File: src/components/f3dGraph/F3DGraph.tsx [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+		update({ nodes: [...filterable(graphState.nodeIndex.values(), (n) => n.type === 'node')], links: [...graphState.linkIndex.values()] });
+		console.log(model);
+	}, [graphPatch]);
+
+	return (
+		<ForceGraph3D
+			width={1000}
+			height={800}
+			graphData={model}
+			enableNodeDrag={true}
```
Comment (DzianisMakeichyk) [inline src/components/f3dGraph/F3DGraph.tsx:96]:
```suggestion
			enableNodeDrag
```

File: src/components/f3dGraph/StoryWrapper.tsx [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+	// 		dispatch({ type: 'connect', payload: { subject: node1, predicate: exampleConnectedPredicate, object: node2 } });
+	// 	}
+	// };
+
+	// const onElementsSelected = (selection: GraphSelection): void => {
+	// 	dispatch({ type: 'updateSelection', payload: selection });
+	// };
+
+	const loadTurtle = (): void => {
```
Comment (DzianisMakeichyk) [inline src/components/f3dGraph/StoryWrapper.tsx:83]: wrapper looks like no needs :)

File: src/components/state/patchGraph.ts [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+const writer = new Writer();
+const quadToString = ({ subject, predicate, object, graph }: Quad) => writer.quadToString(subject, predicate, object, graph);
+
+const initG = () => ({
+	incoming: new Map<string, GraphNode[]>(),
+	outgoing: new Map<string, GraphNode[]>(),
+	properties: new Map<string, string[]>(),
+	links: [] as GraphEdge[],
+});
+const refCount = (node: GraphNode) => [node.incoming.size, node.outgoing.size, node.properties.size, node.links.length].reduce((a, n) => a + n);
```
Comment (DzianisMakeichyk) [inline src/components/state/patchGraph.ts:30]:
```suggestion
const refCount = ({.incoming, outgoing, ..... }: GraphNode) => [node.incoming.size, node.outgoing.size, node.properties.size, node.links.length].reduce((a, n) => a + n);
```

File: src/tests/patchGraph.test.tsx [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+		];
+
+		const patch: RdfPatch2 = quads.map((quad) => {
+			return { action: 'add', assertion: quad };
+		});
+		const graphState = { linkIndex: new Map<string, GraphEdge>(), nodeIndex: new Map<string, GraphNode>() };
+		const res = patchGraph(graphState, patch);
+		const assertions = [...res.graphPatch];
+
+		for (const q of quads) {
```
Comment (DzianisMakeichyk) [inline src/tests/patchGraph.test.tsx:23]:
```suggestion
		for (const { predicate, subject, ........ } of quads) {
```

**Review by DzianisMakeichyk — COMMENTED** [review]
Some visual stuff :)

**Pull request merged**